### PR TITLE
#19 #20 Do not return duplicates from the vector index

### LINKED ISSUE #19: Exclude current document from similarity search
When using the `similar` method on an index, the current document is returned as a similar document. This should be excluded.

### LINKED ISSUE #20: Prevent duplicates in similarity search
When performing a similarity search, duplicate documents may be returned. This method should return a list of unique documents.

### PR DESCRIPTION
**Pull request by tm-kn**
fixes #19
fixes #20

### CODE REVIEW

**Comment by tm-kn** [comment]
@tomusher Tried to write a test but I'm not getting it quite right...

**Review by nimasmi — COMMENTED** [review]
I have added a comment to the deduplication implementation which would change the nature of the way we tested it. With this, you could maybe just test the method directly

```
def test_deduplication():
    assert VectorIndex.deduplicate_list(["a", "b", "c", "b"]) == ["a", "b", "c"]

def test_exclusions():
    assert VectorIndex.deduplicate_list(["a", "b", "c", "d"], exclusions=["c"]) == ["a", "b", "d"]
```

File: src/wagtail_vector_index/index/base.py [hunk 1]
```diff
@@ -58,24 +69,45 @@ def query(self, query: str) -> QueryResponse[VectorIndexableType]:
...
+        # Eliminate duplicates of the same objects.
+        return_list: MutableSequence[VectorIndexableType] = []
+        for return_obj in self.object_type.bulk_from_documents(similar_documents):
+            if not include_self and return_obj == object:
+                continue
+
+            if return_obj in return_list:
+                continue
+            return_list.append(return_obj)
+        return return_list
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply src/wagtail_vector_index/index/base.py:94]: @nimasmi Thanks for this! That does assume that the `dict.fromkeys` preserves order. Is there any harm in that?

[the remark this answers is not in the document]
Comment (nimasmi) [reply src/wagtail_vector_index/index/base.py]: The docs describe it as "Create a new dictionary with keys from _iterable_". Given that in Python 3.7+, dicts preserve the insertion order of their keys, and this method accepts an iterable, it sounds pretty convincing, but it doesn't guarantee that behaviour, no.

File: src/wagtail_vector_index/index/base.py [hunk 2]
```diff
@@ -87,7 +100,24 @@ def search(self, query: str, *, limit: int = 5) -> list[VectorIndexableType]:
...
+    def _deduplicate_list(
+        documents: Iterable[VectorIndexableType],
+        *,
+        exclusions: Iterable[VectorIndexableType] | None = None,
+    ) -> list[VectorIndexableType]:
+        if exclusions is None:
+            exclusions = []
+        # This code assumes that OrderedDict.fromkeys preserves order.
+        return list(
+            OrderedDict.fromkeys(item for item in documents if item not in exclusions)
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply src/wagtail_vector_index/index/base.py]: True! It's a language feature, so other runtimes should be fine. Thanks Nick!

File: tests/test_index.py [hunk 3]
```diff
@@ -99,3 +102,46 @@ def test_index_get_documents_returns_at_least_one_document_per_page():
...
+
+    case = unittest.TestCase()
+
+    # We expect 9 results without the page itself.
+    actual = vector_index.similar(pages[0], limit=100, include_self=False)
+    case.assertCountEqual(actual, pages[1:])
+
+    # We expect 10 results with the page itself.
+    actual = vector_index.similar(pages[0], limit=100, include_self=True)
+    case.assertCountEqual(actual, pages)
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply tests/test_index.py]: That is very neat, thanks Nick, implemented this on the PR!

File: src/wagtail_vector_index/index/base.py [hunk 4]
```diff
@@ -87,7 +101,23 @@ def search(self, query: str, *, limit: int = 5) -> list[VectorIndexableType]:
...
+    @staticmethod
+    def _deduplicate_list(
+        documents: Iterable[VectorIndexableType],
+        *,
+        exclusions: Iterable[VectorIndexableType] | None = None,
+    ) -> list[VectorIndexableType]:
+        if exclusions is None:
+            exclusions = []
+        # This code assumes that dict.fromkeys preserves order which is
+        # behavior of the Python language since version 3.7.
```
Comment (nimasmi) [inline src/wagtail_vector_index/index/base.py:119]: **Comment**: Typing "behavior" is interesting behaviour.

**Review by nimasmi — APPROVED** [review]
Thanks @tm-kn, yes this looks neater now. I approve.

**Pull request merged**